src/main/sensors/esc_sensor.c

```diff
@@ -3307,6 +3308,381 @@ static serialReceiveCallbackPtr graupnerSensorInit(void)
+        uint16_t throttle = buffer[7];
+        uint16_t power = buffer[12];
+        uint16_t voltage = buffer[3] << 8 | buffer[4];
+        uint16_t current = buffer[5] << 8 | buffer[6];
+        uint16_t capacity = buffer[15] << 8 | buffer[16];
+        uint16_t status = buffer[13] << 8 | buffer[14];
+        uint16_t voltBEC = buffer[19];
+
+        escSensorData[0].id = ESC_SIG_XDFLY;
+        escSensorData[0].age = 0;
+        escSensorData[0].erpm = rpm * 10;
+        escSensorData[0].throttle = throttle * 10;
+        escSensorData[0].pwm = power * 10;
+        escSensorData[0].voltage = voltage * 100;
+        escSensorData[0].current = current * 100;
+        escSensorData[0].consumption = capacity;
+        escSensorData[0].temperature = temp * 10;
+        escSensorData[0].bec_voltage = voltBEC * 1000;
+        escSensorData[0].status = status;
+
+        DEBUG(ESC_SENSOR, DEBUG_ESC_1_RPM, rpm * 10);
+        DEBUG(ESC_SENSOR, DEBUG_ESC_1_TEMP, temp * 10);
+        DEBUG(ESC_SENSOR, DEBUG_ESC_1_VOLTAGE, voltage * 10);
+        DEBUG(ESC_SENSOR, DEBUG_ESC_1_CURRENT, current * 10);
+
+        DEBUG(ESC_SENSOR_DATA, DEBUG_DATA_RPM, rpm);
+        DEBUG(ESC_SENSOR_DATA, DEBUG_DATA_PWM, power);
+        DEBUG(ESC_SENSOR_DATA, DEBUG_DATA_TEMP, temp);
+        DEBUG(ESC_SENSOR_DATA, DEBUG_DATA_VOLTAGE, voltage);
+        DEBUG(ESC_SENSOR_DATA, DEBUG_DATA_CURRENT, current);
+        DEBUG(ESC_SENSOR_DATA, DEBUG_DATA_CAPACITY, capacity);
+        DEBUG(ESC_SENSOR_DATA, DEBUG_DATA_EXTRA, status);
+        DEBUG(ESC_SENSOR_DATA, DEBUG_DATA_AGE, 0);
+
+        dataUpdateUs = currentTimeUs;
+
+        totalFrameCount++;
+    }
+    // Maximum frame spacing 50ms, sync after 3 frames
+    checkFrameTimeout(currentTimeUs, 500000);
+}
+
+static void xdflyBuildReq(uint8_t cmd, int *paramIdx, uint16_t framePeriod, uint16_t frameTimeout)
+{
+    uint8_t idx = 0;
+    reqbuffer[idx++] = XDFLY_SYNC;
+    reqbuffer[idx++] = XDFLY_FRAME_LENGTH;
+    reqbuffer[idx++] = cmd;
+    if(cmd == XDFLY_CMD_HANDSHAKE) { // empty payload when sending handshake
+        reqbuffer[idx++] = 0;
+        reqbuffer[idx++] = 0;
+        reqbuffer[idx++] = 0;
+    }else if(cmd == XDFLY_CMD_GET_PARAM) {
+        memcpy(reqbuffer + idx, paramIdx, XDFLY_PAYLOAD_LENGTH);
+        idx++;
+        reqbuffer[idx++] = 0;
+        reqbuffer[idx++] = 0;
+    } else if(cmd == XDFLY_CMD_SET_PARAM) {
+        memcpy(reqbuffer + idx, paramIdx, XDFLY_PAYLOAD_LENGTH);
+        idx++;
+        reqbuffer[idx++] = paramUpdPayload[(*paramIdx)*2+1] | 0x80; //mark as active
+        reqbuffer[idx++] = paramUpdPayload[(*paramIdx)*2];
+    }   
+    uint8_t crclen = idx;
+    uint16_t crc16 = calculateCRC16_MODBUS(reqbuffer, crclen);
+    reqbuffer[idx++] = crc16 >> 8;
+    reqbuffer[idx++] = crc16 & 0xFF;
+
+    reqLength = idx;
+    rrfsmFramePeriod = framePeriod;
+    rrfsmFrameTimeout = frameTimeout;
+}
+
+static void xdflyGetNextParam()
+{
+    if(!xdflyParamCached[xdflyParamIndex]){
```

Comment:
Please follow the general formatting rules in Rotorflight.
Watch out where the spaces go!

Reply:
Sure, will do so, can you point me to the rules?